### ISSUE TITLE
Figure out why userfunc_mandelbrot is so slow

### ISSUE THREAD

**Issue by Enet4** [Enet4/julia-bench-rs#2]
The performance of the [mandelbrot function](https://github.com/Enet4/julia-bench-rs/blob/master/src/main.rs#L64-L90) isn't very well optimized, running even slower than the Java and JavaScript implementations. That is compromising the overall score of Rust alongside the other languages.

The results on my machine:

- C: 1.00
- Julia: 0.80
- Rust: 2.13
- Fortran: 0.77
- Java: 1.25
- JavaScript: 1.30

**Comment by Stargateur** [comment]
To contribute, I win some time by moving `take()` that allow to not perform every time the last expensive computation, and the other problem is that the last `*z = current * current + c` is compute even if `while_take()` will stop the iteration. So remove `while_take()` and use a if into `scan()`, I think it's still idiomatic, and it's win like 3% of time on my computer.

``` rust
fn mandel(z: Complex64) -> u32 {
    use std::iter;

    iter::repeat(z)
        .take(80)
        .scan(z, |z, c| {
            let current = *z;
            if current.norm() > 2.0 {
                None
            } else {
                *z = current * current + c;
                Some(current)
            }
        })
        .count() as u32
}
```

For `mandelperf()`, I maybe found a more idiomatic way to do it:

``` rust
fn mandelperf() -> Vec<u32> {
    (0..21)
        .flat_map(|i| (0..26).map(move |j| (i, j)))
        .map(|(i, j)| {
            mandel(Complex64::new(
                (j as f64 - 20.) / 10.,
                (i as f64 - 10.) / 10.,
            ))
        })
        .collect()
}
```

I think it's `Complex64`, that slow down Mandelbrot benchmark.

**Comment by Enet4** [comment]
@Stargateur Thanks for your feedback. But unfortunately, your suggestions for `mandel` actually seem to slightly slow the benchmark down on my machine (from 0.1614 to 0.1634).

I also experimented with similar improvements in `mandelperf`:

```
fn mandelperf() -> Vec<u32> {
    (-10..=10).flat_map(|i| (-20..=5).map(move |j| (i, j)))
        .map(|(i, j)| (j as f64 / 10., i as f64 / 10.))
        .map(|(re, im)| Complex64::new(re, im))
        .map(mandel)
        .collect()
}
```

This version cuts on the previous `- 20` and `- 10` subtractions, but does not bring any notable improvements.

> I think it's `Complex64`, that slow down Mandelbrot benchmark.

Yes, maybe I should look into the existing methods and see what can be improved. In particular, I just noticed that the Julia version tests for the squared norm of the complex number, which will indeed be faster.

------------

That was it! Replacing `current.norm() <= 2.0` with `current.norm_sqr() <= 4.0` has boosted the bench from 0.161 to 0.063! This is an enormous difference, and does show how we should avoid doing square roots whenever possible :+1: ---I'll be testing against the other benchmarks soon.---

New benchmark score: 0.81, which is almost as fast as Julia's. I'll be pushing these changes now. :+1:
